external/llvm-project/mlir/lib/Conversion/AMDGPUToROCDL/AMDGPUToROCDL.cpp

```diff
@@ -830,6 +830,151 @@ LogicalResult PackedStochRoundFp8OpLowering::matchAndRewrite(
+      Value undefVec = rewriter.create<LLVM::UndefOp>(loc, llvmVecType);
+      src = rewriter.create<LLVM::InsertElementOp>(
+          loc, undefVec, src, createI32Constant(rewriter, loc, 0));
+      src = rewriter.create<LLVM::BitcastOp>(loc, llvmI32Type, src);
+    }
+
+    // This is taken from the following file llvm/lib/Target/AMDGPU/SIDefines.h
+    enum DppCtrl : unsigned {
+      ROW_SHL0 = 0x100,
+      ROW_SHR0 = 0x110,
+      ROW_ROR0 = 0x120,
+      WAVE_SHL1 = 0x130,
+      WAVE_ROL1 = 0x134,
+      WAVE_SHR1 = 0x138,
+      WAVE_ROR1 = 0x13C,
+      ROW_MIRROR = 0x140,
+      ROW_HALF_MIRROR = 0x141,
+      BCAST15 = 0x142,
+      BCAST31 = 0x143,
+    };
+
+    auto kind = DppOp.getKind();
+    auto permArgument = DppOp.getPermArgument();
+    int attrValue = 0;
+
+    switch (kind) {
+
+    case DPPPerm::quad_perm:
+      if (auto quadPermAttr = permArgument->dyn_cast<ArrayAttr>()) {
+        for (int i = 3; i >= 0; --i) {
+          auto intAttr = quadPermAttr[i].dyn_cast<IntegerAttr>();
+          int num = intAttr.getInt();
+          attrValue |= num << (i * 2);
+        }
+      }
+      break;
+    case DPPPerm::row_shl:
+      if (auto intAttr = permArgument->dyn_cast<IntegerAttr>()) {
+        attrValue = intAttr.getInt() + DppCtrl::ROW_SHL0;
+      }
+      break;
+    case DPPPerm::row_shr:
+      if (auto intAttr = permArgument->dyn_cast<IntegerAttr>()) {
+        attrValue = intAttr.getInt() + DppCtrl::ROW_SHR0;
+      }
+      break;
+    case DPPPerm::row_ror:
+      if (auto intAttr = permArgument->dyn_cast<IntegerAttr>()) {
+        attrValue = intAttr.getInt() + DppCtrl::ROW_ROR0;
+      }
+      break;
+    case DPPPerm::wave_shl:
+      attrValue = DppCtrl::WAVE_SHL1;
+      break;
+    case DPPPerm::wave_shr:
+      attrValue = DppCtrl::WAVE_SHR1;
+      break;
+    case DPPPerm::wave_rol:
+      attrValue = DppCtrl::WAVE_ROL1;
+      break;
+    case DPPPerm::wave_ror:
+      attrValue = DppCtrl::WAVE_ROR1;
+      break;
+    case DPPPerm::row_mirror:
+      attrValue = DppCtrl::ROW_MIRROR;
+      break;
+    case DPPPerm::row_half_mirror:
+      attrValue = DppCtrl::ROW_HALF_MIRROR;
+      break;
+    case DPPPerm::row_bcast_15:
+      attrValue = DppCtrl::BCAST15;
+      break;
+    case DPPPerm::row_bcast_31:
+      attrValue = DppCtrl::BCAST31;
+      break;
+    }
+
+    dppCtrl = rewriter.create<LLVM::ConstantOp>(loc, llvmI32Type, attrValue);
+
+    // Check for row_mask, bank_mask, bound_ctrl if they exist and create
+    // constants
+    auto rowMask =
+        createConstantOrUseDefault(DppOp, rewriter, "row_mask", llvmI32Type);
+    auto bankMask =
+        createConstantOrUseDefault(DppOp, rewriter, "bank_mask", llvmI32Type);
+    auto boundCtrl =
+        createConstantOrUseDefault(DppOp, rewriter, "bound_ctrl", llvmI1Type);
+
+    // create a ROCDL_DPPMovOp instruction with the appropriate attributes
+    auto dppMovOp = rewriter.create<ROCDL::DPPMovOp>(
```

Comment:
Can we go clean up the definition of `ROCDL::DPPMovOp` to take those three attributes (row mask, bank mask, and bound control) as I32 attributes instead, to correspond to them being `immarg`s on the LLVM side? It's not critical, but since we're here, we might as well.

Reply:
Did you mean to change instead of I32 to I32Attr in ROCDL_DPPMovOp?
So instead of getting:
rocdl.mov.dpp %arg0 with %0, %1, %2, %3 : i32
getting this at the end:
rocdl.mov.dpp %arg0 with 320, 1, 1, true : i32

Reply:
Yep